src/ObjectUtils/LinuxMap.cpp

```diff
@@ -91,62 +90,102 @@ ErrorMessageOr<std::vector<ModuleInfo>> ReadModules(int32_t pid) {
 
 namespace {
 
-// Loadable sections of an ELF file, including the .text section, are always aligned in the file
-// such that the loader can create a file mapping for them. We can therefore simply detect modules
-// loaded by a process from the executable file mappings.
+// We observed that in some cases, in particular for binaries running under Wine, a single loadable
+// executable segment of an ELF file or a single executable section of a PE can be loaded into
+// memory with multiple adjacent file mappings. In addition, some PEs can have multiple executable
+// sections. Therefore, simply detecting modules loaded by a process from individual executable file
+// mappings won't work.
 //
-// But in the case of Portable Executables, the .text section (and other sections) can have an
-// offset in the file (PointerToRawData, multiple of FileAlignment) that is not congruent to the
-// offset of that section when loaded into memory (VirtualAddress, multiple of SectionAlignment)
-// modulo the page size. This doesn't fulfill the requirements on the arguments of mmap, so in these
-// cases Wine cannot create a file-backed mapping for the .text section, and resorts to creating an
-// anonymous mapping and copying the .text section into it. This means that, for PE binaries with
-// this property, we cannot simply associate an executable mapping to the corresponding file using
-// the path in the mapping.
+// Instead, while scanning the /proc/[pid]/maps file, we can keep track of the module whose mappings
+// we are currently processing. We consider all executable mappings that belong to this module. In
+// the end, we build a ModuleInfo that spans the memory region from the start of the first
+// executable mapping to the end of the last executable mapping for this module.
+//
+// Such ModuleInfo will carry executable_segment_offset with the assumption that the value of
+// ObjectFile::GetExecutableSegmentOffset correspond to the *first* executable mapping. In the
+// normal case of a single executable section and a single executable mapping, ModuleInfo will
+// simply carry the address range of that one mapping.
+//
+// Note that, in the case of multiple executable sections, these are not necessarily adjacent, while
+// ModuleInfo as constructed will represent a single contiguous address range. We believe this is
+// fine, as the address range should still completely belong to the module, even if it can now
+// include non-executable parts, and as the additional complexity of keeping track of multiple
```

Comment:
I believe this can only lead to a problem in case we ever get an incorrect PC that lands in an area that is non-executable but part of the merged region of a module. I'm not sure what happens, I assume we'll try to unwind but will likely fail (read garbage data, won't find unwind info).

If this happens, we must have already made a mistake earlier (to get the incorrect PC), so I don't think this is a huge problem. But it might be worth thinking through what the worst-case scenario here can be.

Reply:
Got it, I missed the right context here. Thanks for clarifying.